sdk/program/src/sysvar/slot_hashes.rs

```diff
@@ -62,6 +69,68 @@ impl Sysvar for SlotHashes {
     }
 }
 
+/// Trait for querying the `SlotHashes` sysvar.
+pub trait SlotHashesSysvar {
+    /// Get a value from the sysvar entries by its key.
+    /// Returns `None` if the key is not found.
+    fn get(key: &Slot) -> Result<Option<Hash>, ProgramError> {
+        get_slot_hash_bytes_with_position(key)
+            .map(|result| result.map(|(_, bytes)| Hash::new(&bytes[8..8 + HASH_BYTES])))
+    }
+
+    /// Get the position of an entry in the sysvar by its key.
+    /// Returns `None` if the key is not found.
+    fn position(key: &Slot) -> Result<Option<usize>, ProgramError> {
+        get_slot_hash_bytes_with_position(key).map(|result| result.map(|(position, _)| position))
+    }
+}
+
+impl SlotHashesSysvar for SlotHashes {}
+
+fn get_slot_hash_bytes_with_position<'a>(
+    slot: &Slot,
+) -> Result<Option<(usize, &'a [u8])>, ProgramError> {
+    // Slot hashes is sorted largest -> smallest slot, so we can leverage
+    // this to perform the search.
+    //
+    // `SlotHashes` can have skipped slots, so we'll have to implement a
+    // binary search over data from `sol_get_sysvar`.
+    let key = slot.to_le_bytes();
+
+    // Rust's `serde::Serialize` will serialize a `usize` as a `u64` on 64-bit
+    // systems for vector length prefixes.
+    let start_offset = std::mem::size_of::<u64>();
+    let length = SlotHashes::size_of().saturating_sub(start_offset);
+    let entry_size = std::mem::size_of::<SlotHash>();
+
+    let data = get_sysvar(&SlotHashes::id(), start_offset as u64, length as u64)?;
```

Comment:
The idea was to not deserialize the data within the BPF context. Previously, programs had no issue loading the whole `SlotHashes`, but as soon as binary search was run on the deserialized type it would bust.

Reply:
You don't need to deserialize the data, you can do an in-place transmute. Check out this little test I whipped up:

```
    #[repr(C)]
    #[derive(Copy, Clone, Debug, Default, Pod, Zeroable)]
    pub struct PodSlotHash {
        slot: Slot,
        hash: Hash,
    }
    impl PartialEq<PodSlotHash> for SlotHash {
        fn eq(&self, other: &PodSlotHash) -> bool {
            self.0 == other.slot && self.1 == other.hash
        }
    }
    #[test]
    fn test_inplace() {
        let mut slot_hashes = SlotHashes::new(&[]);
        for i in 0..MAX_ENTRIES + 1 {
            slot_hashes.add(
                i as u64,
                hash(&[(i >> 24) as u8, (i >> 16) as u8, (i >> 8) as u8, i as u8]),                   );
        }
        let data = bincode::serialize(&slot_hashes).unwrap();
        let pod_hashes: &[PodSlotHash] = bytemuck::try_cast_slice(&data[8..]).unwrap();
        assert_eq!(&slot_hashes.0, pod_hashes);

        let search_slot = 11u64;
        // position()
        assert_eq!(
            pod_hashes.binary_search_by(|PodSlotHash { slot, .. }| search_slot.cmp(slot))
                .unwrap(),
            slot_hashes.position(&search_slot).unwrap(),
        );
        // get()
        assert_eq!(
            pod_hashes
                .binary_search_by(|PodSlotHash { slot, .. }| search_slot.cmp(slot))
                .map(|index| &pod_hashes[index].hash)
                .unwrap(),
            slot_hashes.get(&search_slot).unwrap(),
        );
    }
```

Reply:
Ahh interesting. Very slick! I would need to introduce this struct in order to implement `Pod`, etc. on the entry types though, right? Afaict you can't do it on just a type alias.

Reply:
Yeah exactly, it'll require a new struct unfortunately. And if you use a plain `u64` as in my test, you'll need to make sure that the address you use is aligned to 8

Reply:
Luckily bincode uses a `u64` for vector length and the has is 32 bytes, so alignment should be fairly trivial, no?

Reply:
Unless I'm missing something, the syscall doesn't check for any alignment on the given address, so for example, if you provide the address `0x300000001` to the "get sysvar" syscall, then it will write the data starting at that address, and it won't be properly aligned to 8

Reply:
Ok cool I've added a commit using pod types. I think I still need to add some alignment validation.

Syscalls do have alignment checking when translating types to given addresses, but I haven't dug into it yet to see if this addresses your concern.

https://github.com/anza-xyz/agave/blob/e0dc5dc707ab80ac469e6d537244e083811e3e85/programs/bpf_loader/src/syscalls/sysvar.rs#L6

Reply:
I've added an alignment check to the API regardless!